ob1k-consul/src/main/java/com/outbrain/ob1k/consul/CooperativeTargetProvider.java

```diff
@@ -0,0 +1,126 @@
+package com.outbrain.ob1k.consul;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ConcurrentHashMultiset;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multiset;
+import com.google.common.collect.Ordering;
+import com.outbrain.ob1k.client.targets.TargetProvider;
+
+/**
+ * A {@link TargetProvider} that chooses targets likely to respond fast and successfully.
+ *
+ * @author Amir Hadadi
+ */
+public class CooperativeTargetProvider extends ConsulBasedTargetProvider {
+
+  private static final Logger log = LoggerFactory.getLogger(CooperativeTargetProvider.class);
+  private static final int EXTRA_TARGETS = 10;
+  private final int penalty;
+  private final Multiset<String> pendingRequests = ConcurrentHashMultiset.create();
+  private final Multiset<String> remainingPenalty = ConcurrentHashMultiset.create();
+  private final AtomicInteger roundRobin = new AtomicInteger();
```

Comment:
The reason we have a threadlocal in the original implementation is to avoid contention.
This implementation contains many contention points, and I doubt if it will perform nicely under load.